src/openzaak/utils/migrations.py

```diff
@@ -0,0 +1,68 @@
+# SPDX-License-Identifier: EUPL-1.2
+# Copyright (C) 2022 Dimpact
+import logging
+from urllib.parse import urlsplit, urlunsplit
+
+from django.db.models.functions import Length
+
+logger = logging.getLogger(__name__)
+
+
+def get_service(model, url: str):
+    """
+    copy-paste from zgw_consumers.Service.get_service method, which can't be used
+    during migrations
+    """
+    split_url = urlsplit(url)
+    scheme_and_domain = urlunsplit(split_url[:2] + ("", "", ""))
+
+    candidates = (
+        model.objects.filter(api_root__startswith=scheme_and_domain)
+        .annotate(api_root_length=Length("api_root"))
+        .order_by("-api_root_length")
+    )
+
+    # select the one matching
+    for candidate in candidates.iterator():
+        if url.startswith(candidate.api_root):
+            return candidate
+
+    return None
+
+
+def fill_service_urls(
+    apps,
+    model,
+    url_field: str,
+    service_base_field: str,
+    service_relative_field: str,
+    fake_etag: bool = False,
+):
+    """
+    helper function to migrate from UrlField to ServiceUrlField
+
+    :param fake_etag: workaround to migrate ETagMixin models
+    """
+    Service = apps.get_model("zgw_consumers", "Service")
+    cache_get_service = {}
+
+    for instance in model.objects.exclude(**{url_field: ""}):
+        url = getattr(instance, url_field)
+
+        if url in cache_get_service:
+            service = cache_get_service[url]
+        else:
+            service = get_service(Service, url)
```

Comment:
this does not handle the case/situation where the service doesn't exist (anymore) for the URL, and will then lead to migration crashes.

I'd do a detection loop first and create the missing services with a FIXME label based on a heuristic (you should be able to detected `/api/v1` in the URL and treat the prefix + that part as the root) to avoid this kind of breakage.

I'd also like to see some tests for these data migrations/utilities with those edge cases.

Reply:
I will add the creation of missing services with FIXME label  and warning in logs but tbh I don't know how to test it. 
This function uses the URLField in the model which is already deleted
And I don't want to create some model just for testing, and leave it sitting in DB.

Reply:
you can use this: https://github.com/open-formulieren/open-forms/blob/master/src/openforms/utils/tests/test_migrations.py#L8 :)

Reply:
Tests are added, thanks for the link